nodejs/src/stream_writer.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright Pravega Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use neon::prelude::*;
+use neon::types::buffer::TypedArray;
+use pravega_client::error::Error as WriterError;
+use pravega_client::event::writer::EventWriter;
+use pravega_client::util::oneshot_holder::OneShotHolder;
+use pravega_client_shared::ScopedStream;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::runtime::Handle;
+use tokio::sync::oneshot::Receiver;
+use tokio::sync::Mutex;
+use tokio::time::timeout;
+
+impl Finalize for StreamWriter {
+    // TODO: impl finalize for this so events will be flushed
```

Comment:
Yes! It is left undone because implementing other features are more important at that commit. I add the `finalize` method and can be verified by following test snippets.

```rust
impl Finalize for StreamWriter {
    fn finalize<'a, C: Context<'a>>(self, cx: &mut C) {
        let _ = js_log(cx, "gc calls flush".to_string());
        self.runtime_handle.block_on(async {
            let _ = self.writer.lock().await.flush().await;
        });
    }
}
```

```typescript
// omit other code in client.test.ts
await (async () => {
    const stream_writer_1 = stream_manager.create_writer(SCOPE, STREAM);
    await stream_writer_1.write_event(DATA);
    await stream_writer_1.write_event(DATA, 'routing_key');
    const enc = new TextEncoder();
    const stream_writer_2 = stream_manager.create_writer(SCOPE, STREAM);
    stream_writer_2.write_event_bytes(enc.encode(DATA));
    stream_writer_2.write_event_bytes(enc.encode(DATA), 'routing_key');
    await stream_writer_2.flush();
    })();
global.gc();
// omit other code in client.test.ts
```

```bash
$ PRAVEGA_NODEJS_DEV=true node --expose-gc --loader ts-node/esm --experimental-vm-modules node_modules/jest/bin/jest.js --runInBand client.test.ts
(node:18558) ExperimentalWarning: --experimental-loader is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
(node:18558) ExperimentalWarning: VM Modules is an experimental feature. This feature could change at any time
gc calls flush    <-- here
gc calls flush    <-- duplicate because we have 2 writers
 PASS  tests/client.test.ts
  Basic test on manager, reader, and writer
    ✓ main (828 ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        2.162 s
Ran all test suites matching /client.test.ts/i.
```

Reply:
However, the tests committed will not contain this, it is the bindings job to make sure `finalize` is called on gc and rust client job to make sure events are flushed.

Reply:
I also tested the behavior when Pravega is offline before gc, the `flush()` call in `finalize` won't block the process.